### PR TITLE
Allow to use Twig inside block content + create some shortcut function

### LINKED ISSUE #33: Render block content as twig
If we render block content as twig it will give the end user enormous power. Probably we should execute it in a twig sandbox environment, but still it could allow to execute functions like:

- `{{ render_taxon('code', 'template') }}`
- `{{ render_product('code', 'template') }}`
- `{{ render_xxx('code', 'template') }}`

### PR DESCRIPTION
**Pull request by Roshyo**
Fixes #33

### CODE REVIEW

File: src/Twig/Extension/Runtime.php [hunk 1]
```diff
@@ -29,4 +46,95 @@ public function view(string $view): string
     {
         return $this->viewRenderer->render($view);
     }
+
+    /**
+     * @throws RouteNotFoundException              If the named route doesn't exist
```
Comment (loevgaard) [inline src/Twig/Extension/Runtime.php:52]: Should we catch these exceptions and add some html comments to explain the situation to the marketing guy doing this? :D

File: src/Twig/Extension/Runtime.php [hunk 2]
```diff
@@ -29,4 +46,95 @@ public function view(string $view): string
...
+            'sylius.product',
+            $displayedValue,
+            ['slug' => $slug, '_locale' => $localeCode],
+            $referenceType
+        );
+    }
+
+    public function linkToTaxon(
+        string $slug,
+        ?string $displayedValue = null,
```
Comment (loevgaard) [inline src/Twig/Extension/Runtime.php]:
```suggestion
        string $displayedValue = null,
```

File: src/Twig/Extension/Runtime.php [hunk 3]
```diff
@@ -29,4 +46,95 @@ public function view(string $view): string
...
+        ?string $localeCode = null,
```
Comment (loevgaard) [inline src/Twig/Extension/Runtime.php]:
```suggestion
        string $localeCode = null,
```

File: src/Twig/Extension/Runtime.php (hunk 2 above)
Comment (loevgaard) [reply src/Twig/Extension/Runtime.php]: Make this change all the applicable places

File: src/Twig/Extension/Runtime.php (hunk 1 above)
Comment (Roshyo) [reply src/Twig/Extension/Runtime.php:52]: Yes, and in an ideal world, we should implement a validator on submit that would parse those and show exceptions.

Or perhaps the validator can even try to render, and catch exceptions :thinking:

File: src/Renderer/BlockRenderer.php [hunk 4]
```diff
@@ -52,9 +52,15 @@ public function render($block): string
 
         return $this->twig->render('@SetonoSyliusCMSPlugin/block.html.twig', [
             'block' => $block,
+            'renderedBlock' => $this->renderBlockContent($block->getContent() ?? ''),
```
Comment (loevgaard) [inline src/Renderer/BlockRenderer.php]:
```suggestion
            'renderedBlock' => $block->getContent() ? $this->renderBlockContent($block->getContent()) : '',
```
Isn't this a bit cleaner?

Comment (Roshyo) [reply src/Renderer/BlockRenderer.php]: I think the current code is clean enough. Perhaps you're just more used to the other one

File: src/Twig/Extension/Runtime.php [hunk 5]
```diff
@@ -29,4 +47,99 @@ public function view(string $view): string
...
+        string $slug,
+        string $displayedValue = null,
+        string $localeCode = null,
+        int $referenceType = UrlGeneratorInterface::ABSOLUTE_PATH
+    ): string {
+        if (null === $localeCode) {
+            $localeCode = $this->localeContext->getLocaleCode();
+        }
+
+        return $this->linkToResource(
```
Comment (loevgaard) [inline src/Twig/Extension/Runtime.php:101]: Not sure why you don't use the `linkToRoute` here

**Review by loevgaard — APPROVED** [review]
Nice PR 🎉

File: src/Twig/Extension/Runtime.php (hunk 5 above)
Comment (Roshyo) [reply src/Twig/Extension/Runtime.php:101]: We can also. Both would be correct

File: src/Renderer/BlockRenderer.php [hunk 6]
```diff
@@ -50,11 +51,27 @@ public function render($block): string
...
         return $this->twig->render('@SetonoSyliusCMSPlugin/block.html.twig', [
             'block' => $block,
+            'renderedBlock' => $renderedBlockContent,
         ]);
     }
 
+    /**
+     * @throws Error
+     */
+    private function renderBlockContent(string $blockContent): string
```
Comment (loevgaard) [inline src/Renderer/BlockRenderer.php:70]: Given the name of this method I would probably have the try catch inside this one, but I guess that's taste

**Review by loevgaard — APPROVED** [review]

**Pull request merged**